### PR TITLE
Add coverage for smart class parameters functionality (1 phase)

### PR DESCRIPTION
**Pull request by oshtaier**
```
nosetests tests/foreman/cli/test_classparameters.py
SSSS..S........SSSSSS...SSSSS.SS..........
----------------------------------------------------------------------
Ran 42 tests in 284.742s

OK (SKIP=18)
```

Blocked by #3575

### CODE REVIEW

File: tests/foreman/cli/test_scparam.py [hunk 1]
```diff
@@ -1,47 +0,0 @@
-# -*- encoding: utf-8 -*-
-"""Test class for Smart Class Parameter CLI.
```
Comment (oshtaier) [inline tests/foreman/cli/test_scparam.py]: That module is duplication of `test_classparameters.py`, so removing in favor of it, because it contains much more test cases

File: tests/foreman/cli/test_scparam.py [hunk 2]
```diff
@@ -1,47 +0,0 @@
...
-    def run_puppet_module(self):
-        """Import some parameterized puppet class. This is required to make
-        sure that we have smart class variable available.
-
-        """
-        ssh.command('puppet module install --force puppetlabs/ntp')
-
-    @run_only_on('sat')
-    @tier1
-    def test_positive_list(self):
```
Comment (oshtaier) [inline tests/foreman/cli/test_scparam.py]: Removing test as it is already covered in `setUpClass` method

File: robottelo/cli/base.py [hunk 3]
```diff
@@ -443,7 +440,7 @@ def _construct_command(cls, options=None):
             elif val is not False:
                 if isinstance(val, list):
                     val = ','.join(str(el) for el in val)
-                tail += u' --{0}="{1}"'.format(key, val)
+                tail += u" --{0}='{1}'".format(key, val)
```
Comment (oshtaier) [inline robottelo/cli/base.py]: It doesn't seem as impacting any tests, but failing for current suite, so after some re-tests decided to use that idea

hammer -v -u admin -p changeme --output=csv sc-param update --override="1" --id="3" --default-value="${$ntp::params::config_test}"
-bash: ${$ntp::params::config_test}: bad substitution

hammer -v -u admin -p changeme --output=csv sc-param update --override="1" --id="3" --default-value='${$ntp::params::config_test}'
Message,Id
Parameter updated,3

**Comment by oshtaier** [comment]
```
nosetests tests/foreman/cli/test_classparameters.py
SSSS..........SSSSSS........S...........
----------------------------------------------------------------------
Ran 40 tests in 725.838s

OK (SKIP=11)
```

**Comment by sghai** [comment]
Awesome coverage and looks good to me.. thanks @oshtaier 

@jyejare : Since you tested this feature in upstream, so would be great if you can take a look and see if all scenarios are covered. If not, then please file another issue with new scenarios. thanks

File: robottelo/cli/factory.py [hunk 4]
```diff
@@ -1514,7 +1516,8 @@ def make_hostgroup(options=None):
         u'puppet-ca-proxy-id': None,
         u'puppet-proxy': None,
         u'puppet-proxy-id': None,
-        u'puppetclass-ids': None,
+        u'puppet-class-ids': None,
```
Comment (jyejare) [inline robottelo/cli/factory.py]: Follow alphabetical order.

File: robottelo/cli/factory.py [hunk 5]
```diff
@@ -1514,7 +1516,8 @@ def make_hostgroup(options=None):
...
+        u'puppet-classes': None,
```
Comment (jyejare) [inline robottelo/cli/factory.py]: Follow alphabetical order.

**Comment by jyejare** [comment]
Please rename smartclass.py to (smartclassparameters/smartparameters/scparams).py

File: robottelo/cli/smartclass.py [hunk 6]
```diff
@@ -11,17 +11,76 @@
 
 Subcommands::
 
-    info                          Show a smart class parameter.
+    add-override-value            Create an override value for a specific smart
+                                  variable
```
Comment (jyejare) [inline robottelo/cli/smartclass.py]: Should be parameter and not variable.

File: robottelo/cli/smartclass.py [hunk 7]
```diff
@@ -11,17 +11,76 @@
...
+    info                          Show a smart class parameter
     list                          List all smart class parameters
-    update                        Update a smart class parameter.
+    remove-override-value         Delete an override value for a specific smart
+                                  variable
```
Comment (jyejare) [inline robottelo/cli/smartclass.py]: Should be parameter and not variable.

File: robottelo/cli/smartclass.py (hunk 6 above)
Comment (oshtaier) [inline robottelo/cli/smartclass.py]: I understand, but here we put just copy from official documentation

File: robottelo/cli/smartclass.py (hunk 7 above)
Comment (oshtaier) [inline robottelo/cli/smartclass.py]: Same here

File: robottelo/cli/smartclass.py [hunk 8]
```diff
@@ -11,17 +11,76 @@
...
+                                                                yes/no, 1/0.
+            --value VALUE                                       Override value
+        """
+        cls.command_sub = 'add-override-value'
+        return cls.execute(
+            cls._construct_command(options), output_format='csv')
+
+    @classmethod
+    def remove_override_value(cls, options=None):
+        """Delete an override value for a specific smart variable
```
Comment (jyejare) [inline robottelo/cli/smartclass.py]: You mean to say 'parameter' here ?

File: tests/foreman/cli/test_classparameters.py [hunk 9]
```diff
@@ -16,107 +16,253 @@
...
+    skip_if_bug_open,
+    stubbed,
+    tier1,
+    tier2,
+)
 from robottelo.test import CLITestCase
 
 
+@filtered_datapoint
+def valid_sc_parameters():
```
Comment (jyejare) [inline tests/foreman/cli/test_classparameters.py]: s/valid_sc_parameters/valid_scparam_types_values ?
or
s/valid_sc_parameters/valid_scparam_valuetypes ?

File: tests/foreman/cli/test_classparameters.py [hunk 10]
```diff
@@ -16,107 +16,253 @@
...
+    """Returns a list of valid smart class parameter types and values"""
+    return [
+        {
+            u'sc_type': 'string',
+            u'value': gen_string('utf8'),
```
Comment (jyejare) [inline tests/foreman/cli/test_classparameters.py]: Any specific reason of having 'utf8' only ?

File: tests/foreman/cli/test_classparameters.py [hunk 11]
```diff
@@ -16,107 +16,253 @@
...
+            u'sc_type': 'boolean',
+            u'value': choice(['0', '1']),
+        },
+        {
+            u'sc_type': 'integer',
+            u'value': gen_string('numeric', 5),
+        },
+        {
+            u'sc_type': 'real',
+            u'value': -123.0,
```
Comment (jyejare) [inline tests/foreman/cli/test_classparameters.py]: Instead of giving a hardcoded value, its better to have real value generator.

File: tests/foreman/cli/test_classparameters.py [hunk 12]
```diff
@@ -16,107 +16,253 @@
...
+        },
+        {
+            u'sc_type': 'array',
+            u'value': "['ZvqqjmtkeW', 'SGnfs123yekPk']",
```
Comment (jyejare) [inline tests/foreman/cli/test_classparameters.py]: same here as real value.

File: tests/foreman/cli/test_classparameters.py [hunk 13]
```diff
@@ -16,107 +16,253 @@
...
+        },
+        {
+            u'sc_type': 'hash',
+            u'value': {'a': 'test'},
```
Comment (jyejare) [inline tests/foreman/cli/test_classparameters.py]: same as array.

File: tests/foreman/cli/test_classparameters.py [hunk 14]
```diff
@@ -16,107 +16,253 @@
...
+        },
+        {
+            u'sc_type': 'yaml',
+            u'value': 'name=>XYZ',
```
Comment (jyejare) [inline tests/foreman/cli/test_classparameters.py]: better to define a constant/permanent file having yaml data or somewhere in constants.py.

File: tests/foreman/cli/test_classparameters.py [hunk 15]
```diff
@@ -16,107 +16,253 @@
...
+        },
+        {
+            u'sc_type': 'json',
+            u'value': '{\\"name\\": \\"XYZ\\"}',
```
Comment (jyejare) [inline tests/foreman/cli/test_classparameters.py]: same as yaml.

File: tests/foreman/cli/test_classparameters.py [hunk 16]
```diff
@@ -194,12 +355,38 @@ def test_positive_update_parameter_type(self):
...
+                })
+                if data['sc_type'] == 'boolean':
+                    self.assertEqual(
+                        sc_param['default-value'],
+                        True if data['value'] == '1' else False
+                    )
+                elif data['sc_type'] == 'array':
+                    string_list = [
+                        str(element) for element in sc_param['default-value']]
+                    self.assertEqual(str(string_list), data['value'])
```
Comment (jyejare) [inline tests/foreman/cli/test_classparameters.py]: Converting array to string would cause problem here, I believe.

Comment (oshtaier) [inline tests/foreman/cli/test_classparameters.py]: Nope, we converted each element of array (from unicode to string) first and then here converted array as a whole to one string and compare with expected string

File: tests/foreman/cli/test_classparameters.py [hunk 17]
```diff
@@ -219,56 +406,26 @@ def test_negative_update_parameter_type(self):
...
+                        'default-value': test_data['value'],
+                        'id': sc_param_id,
+                        'override': 1,
+                    })
+                sc_param = SmartClassParameter.info({
+                    'puppet-class': 'ntp',
+                    'id': sc_param_id,
+                })
+                self.assertNotEqual(
+                    sc_param['default-value'], test_data['value'])
```
Comment (jyejare) [inline tests/foreman/cli/test_classparameters.py]: I think comparison is not happening properly here in case of boolean, array, json and yaml. We will miss an actual failure here. Good to follow the 'test_positive_update_parameter_type' testcase approach.

Comment (oshtaier) [inline tests/foreman/cli/test_classparameters.py]: It's negative scenario, so I am verifying that exception is raised and that value was not changed, so it doesn't make sense to have such complex logic that we have for positive scenario as we don't care about the content, only about the fact whether is something changed or not

File: tests/foreman/cli/test_classparameters.py [hunk 18]
```diff
@@ -282,38 +439,24 @@ def test_positive_validate_default_value_required_check(self):
...
-        @caseautomation: notautomated
-
-        @CaseLevel: System
         """
+        sc_param_id = self.sc_params_ids_list.pop()
+        SmartClassParameter.update({
+            'parameter-type': 'boolean',
+            'id': sc_param_id,
+            'default-value': u'true',
+            'override': 1,
```
Comment (jyejare) [inline tests/foreman/cli/test_classparameters.py]: When you say only override and 
1. Dont provide any match and no value then the error is 'Match is invalid and Value can't be blank ', which is not the intention of this TC.
but,
2. Provide a match and no value then the error is 'Value can't be blank ', which is the intention of this TC.
And so we will miss the failure here. So please please add a match but no value.

File: tests/foreman/cli/test_classparameters.py [hunk 19]
```diff
@@ -282,38 +439,24 @@ def test_positive_validate_default_value_required_check(self):
...
+            'required': 1
+        })
+        sc_param = SmartClassParameter.info({
+            'puppet-class': 'ntp',
+            'id': sc_param_id,
+        })
+        self.assertEqual(sc_param['required'], True)
+        self.assertEqual(sc_param['default-value'], True)
```
Comment (jyejare) [inline tests/foreman/cli/test_classparameters.py]: Here instead of checking the default value is set, we should check the matcher value is not set and that is the assert statement says.

File: tests/foreman/cli/test_classparameters.py [hunk 20]
```diff
@@ -282,38 +439,24 @@ def test_positive_validate_default_value_required_check(self):
...
-
-        @assert: Error raised for blank matcher value.
-
-        @caseautomation: notautomated
-
-        @CaseLevel: System
         """
+        sc_param_id = self.sc_params_ids_list.pop()
+        SmartClassParameter.update({
+            'parameter-type': 'boolean',
```
Comment (jyejare) [inline tests/foreman/cli/test_classparameters.py]: I found a bug here for parameter type boolean in satellite snap 19.1, that

if we set boolean param type, check 'required' checkbox and no value for matcher, then instead throwing 'value cant be blank' error, the submit is happening with value 'false' by default.

File: tests/foreman/cli/test_classparameters.py (hunk 18 above)
Comment (oshtaier) [inline tests/foreman/cli/test_classparameters.py]: Steps:
```
         1.  Override the parameter.
         2.  Provide some default value, Not empty.
         3.  Set '--required' check.
          4.  Submit the change.
```
Probably, you looked into another case

Case that you mentioned is impossible to verify in CLI

`'Match is invalid and Value can't be blank '` and `'Value can't be blank '` doesn't exist in CLI. Please, don't mess with UI

File: tests/foreman/cli/test_classparameters.py (hunk 19 above)
Comment (oshtaier) [inline tests/foreman/cli/test_classparameters.py]: Replied above

File: tests/foreman/cli/test_classparameters.py (hunk 20 above)
Comment (oshtaier) [inline tests/foreman/cli/test_classparameters.py]: Don't mess with UI, CLI doesn't care about this details

Comment (oshtaier) [inline tests/foreman/cli/test_classparameters.py]: BTW, 'required' checkbox has nothing in common with SC params itself, but more with 'hosts'. Check that bug out for more details:
https://bugzilla.redhat.com/show_bug.cgi?id=1352491
Just to make sure that you understand logic in a right way.
Even in UI, we need to override parameter on the host page to see how required flag is working

**Comment by oshtaier** [comment]
```
nosetests tests/foreman/cli/test_classparameters.py -m test_positive_update_parameter_type
.
----------------------------------------------------------------------
Ran 1 test in 76.201s

OK
```
```
nosetests tests/foreman/cli/test_classparameters.py -m test_positive_validate_matcher_value_required_check
.
----------------------------------------------------------------------
Ran 1 test in 31.090s

OK
```

File: tests/foreman/cli/test_classparameters.py (hunk 10 above)
Comment (oshtaier) [inline tests/foreman/cli/test_classparameters.py]: Should be addressed in https://github.com/SatelliteQE/robottelo/issues/3620

File: tests/foreman/cli/test_classparameters.py (hunk 11 above)
Comment (oshtaier) [inline tests/foreman/cli/test_classparameters.py]: Should be addressed in https://github.com/SatelliteQE/robottelo/issues/3620

File: tests/foreman/cli/test_classparameters.py (hunk 12 above)
Comment (oshtaier) [inline tests/foreman/cli/test_classparameters.py]: Should be addressed in https://github.com/SatelliteQE/robottelo/issues/3620

File: tests/foreman/cli/test_classparameters.py (hunk 15 above)
Comment (oshtaier) [inline tests/foreman/cli/test_classparameters.py]: Should be addressed in https://github.com/SatelliteQE/robottelo/issues/3620

**Comment by oshtaier** [comment]
If there is any comment regarding test logic in test cases, please provide specific hammer commands and outputs to check it

**Comment by oshtaier** [comment]
Addressed all comments so far

**Comment by sghai** [comment]
ACK

File: tests/foreman/cli/test_classparameters.py (hunk 19 above)
Comment (jyejare) [inline tests/foreman/cli/test_classparameters.py]: I think providing no value at least for matcher while 'required' check should raise a bug regardless of UI/CLI.
The functionality for feature should be identical across UI/CLI/API.

File: tests/foreman/cli/test_classparameters.py (hunk 18 above)
Comment (jyejare) [inline tests/foreman/cli/test_classparameters.py]: dear @oshtaier ,
This does exist in CLI as below:
```
[root@qeblade36 ~]# hammer sc-param add-override-value --match os=rhel6 --value '' --smart-class-parameter-id 331
Could not create the override_value:
  Validation failed: Value can't be blank
```

File: tests/foreman/cli/test_classparameters.py (hunk 20 above)
Comment (jyejare) [inline tests/foreman/cli/test_classparameters.py]: As confirmed by dev, the requied checkbox wont raise any error for 'default value' but should raise for matcher/override.

File: tests/foreman/cli/test_classparameters.py (hunk 19 above)
Comment (jyejare) [inline tests/foreman/cli/test_classparameters.py]: Sorry for considering wrong assert statement here! I am ok with asserts.

File: tests/foreman/cli/test_classparameters.py [hunk 21]
```diff
@@ -328,15 +471,26 @@ def test_positive_validate_matcher_value_required_check(self):
...
+        SmartClassParameter.update({
+            'id': sc_param_id,
+            'override': 1,
+            'required': 1
+        })
+        sc_param = SmartClassParameter.info({
+            'puppet-class': 'ntp',
+            'id': sc_param_id,
+        })
+        self.assertEqual(sc_param['required'], True)
```
Comment (jyejare) [inline tests/foreman/cli/test_classparameters.py]: assert if the matcher value is set as given. It lists all overrides in hammer 'info'.

**Comment by oshtaier** [comment]
Added one more test case for matcher with empty value:
```
nosetests tests/foreman/cli/test_classparameters.py -m test_negative_create_matcher
.SSSS
----------------------------------------------------------------------
Ran 5 tests in 53.447s

OK (SKIP=4)
```

File: tests/foreman/cli/test_classparameters.py [hunk 22]
```diff
@@ -948,55 +1292,43 @@ def test_negative_create_matcher_avoid_duplicate(self):
...
+        sc_param = SmartClassParameter.info({
+            'puppet-class': 'ntp',
+            'id': sc_param_id,
+        })
+        self.assertEqual(len(sc_param['override-values']['values']), 0)
 
     @run_only_on('sat')
-    @stubbed()
-    @tier3
-    def test_positive_impact_parameter_delete_attribute(self):
```
Comment (jyejare) [inline tests/foreman/cli/test_classparameters.py]: Why this is been deleted ?

File: tests/foreman/cli/test_classparameters.py [hunk 23]
```diff
@@ -16,107 +16,253 @@
...
 
-        @caseautomation: notautomated
-
-        @CaseLevel: System
-        """
-
-    @run_only_on('sat')
-    @stubbed()
-    @tier3
-    def test_positive_list_parameters_by_puppetclass_name(self):
```
Comment (jyejare) [inline tests/foreman/cli/test_classparameters.py]: Why deleted ?

File: tests/foreman/cli/test_classparameters.py [hunk 24]
```diff
@@ -16,107 +16,253 @@
...
-
-        @caseautomation: notautomated
-
-        @CaseLevel: System
-        """
-
-    @run_only_on('sat')
-    @stubbed()
-    @tier3
-    def test_positive_list_parameters_by_puppetclass_id(self):
```
Comment (jyejare) [inline tests/foreman/cli/test_classparameters.py]: Why deleted ?

**Comment by jyejare** [comment]
ACK, pending why some TC's deleted.

File: tests/foreman/cli/test_classparameters.py (hunk 23 above)
Comment (oshtaier) [inline tests/foreman/cli/test_classparameters.py]: Because, they were not actual, because it was not possible to use puppet class for list command, but I can, so will return these two test cases in a minute.
Of course, they are covered explicitly in setup class, but let's add them anyway

File: tests/foreman/cli/test_classparameters.py (hunk 22 above)
Comment (oshtaier) [inline tests/foreman/cli/test_classparameters.py]: CLI does not operate with thing like attribute. You do not need to create new attribute like 'is_virtual', to create new matcher `is_virtual=true` like we do in UI. Here you just write `is_virtual=true` and get the result, so it is impossible to delete something which was not created

**Comment by oshtaier** [comment]
```
nosetests tests/foreman/cli/test_classparameters.py -m test_positive_list_parameters
......
----------------------------------------------------------------------
Ran 6 tests in 98.721s

OK
```

**Comment by jyejare** [comment]
Lets add required check negative scenarios in phase 2 of implementation as we are stuck with the bug: 
https://bugzilla.redhat.com/show_bug.cgi?id=1352491


And Lets merge this !

**Comment by jyejare** [comment]
ACK
